### PR TITLE
Added support for Google Cloud DNS API (dns_gcloud)

### PR DESCRIPTION
**Pull request by lenartj**
See also #180

### CODE REVIEW

**Comment by Neilpang** [comment]
please move it the last position.

**Comment by thejaswip** [comment]
I can see this is not yet merged into dev branch but marked as closed

File: dnsapi/dns_gcloud.sh [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+    _debug tr "$(cat "$tr")"
+    rm -r "$trd"
+    _err "_dns_gcloud_execute_tr: failed to execute transaction"
+    return 1
+  fi
+  rm -r "$trd"
+
+  for i in $(seq 1 120); do
+    if gcloud dns record-sets changes list \
+      --zone=lenart \
```
Comment (chessmango) [inline dnsapi/dns_gcloud.sh]: Bit hardcoded here xD

Comment (jkleckner) [reply dnsapi/dns_gcloud.sh]: Risking being a little more directive than @chessmango , I believe he means you should replace `lenart` with `"$managedZone"`.  ;-)

You might also indicate what sort of testing you've done to speed up acceptance.

**Comment by jkleckner** [comment]
@lenartj The changes above look reasonably small - any chance to re-submit?

File: dnsapi/dns_gcloud.sh [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+    fi
+  done
+
+  _err "_dns_gcloud_execute_tr: transaction is still pending after 10 minutes"
+  rm -r "$trd"
+  return 1
+}
+
+_dns_gcloud_remove_rrs() {
+  if ! xargs --no-run-if-empty gcloud dns record-sets transaction remove \
```
[the remark this answers is not in the document]
Comment (lenartj) [reply dnsapi/dns_gcloud.sh]: There might be more than one value to be removed (a TXT can be multivalue), so we need the xargs.

File: dnsapi/dns_gcloud.sh [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+  _debug filter "$filter"
+
+  # List domains and find the longest match (in case of some levels of delegation)
+  if ! match=$(gcloud dns managed-zones list \
+    --format="value(name, dnsName)" \
+    --filter="$filter" \
+    | while read -r dnsName name; do
+      printf "%s\t%s\t%s\n" "${#dnsName}" "$dnsName" "$name"
+    done \
+    | sort -n -r | head -n1 | cut -f2,3 | grep '^.*'); then
```
[the remark this answers is not in the document]
Comment (lenartj) [reply dnsapi/dns_gcloud.sh]: done

File: dnsapi/dns_gcloud.sh (hunk 1 above)
Comment (lenartj) [reply dnsapi/dns_gcloud.sh]: whops, fixed. Worked for me! :-D

**Comment by lenartj** [comment]
I've replied to your comments and rebased the code.

About the [guide](https://github.com/Neilpang/acme.sh/wiki/DNS-API-Dev-Guide), I think 6/2 is not practical for gcloud; about 6/5 I think I have grep, cut, head, xargs that are not shell builtins (and printf, depending on the shell). It is possible to rewrite the code not use xargs and printf, but I do not have the time for it, sorry.

**Comment by Neilpang** [comment]
please merge the latest code.

**Comment by lenartj** [comment]
Done :-)

**Pull request merged**